test/Interop/Cxx/class/constructors-ir.swift

```diff
@@ -0,0 +1,42 @@
+// Target-specific tests for C++ constructor call code generation.
+
+// RUN: %swift -module-name Swift -target x86_64-apple-macosx10.9 -dump-clang-diagnostics -I %S/Inputs -enable-cxx-interop -emit-ir %s -parse-stdlib -parse-as-library -disable-legacy-type-info | %FileCheck %s -check-prefix=ITANIUM_X64
+// RUN: %swift -module-name Swift -target armv7-none-linux-androideabi -dump-clang-diagnostics -I %S/Inputs -enable-cxx-interop -emit-ir %s -parse-stdlib -parse-as-library -disable-legacy-type-info | %FileCheck %s -check-prefix=ITANIUM_ARM
+// RUN: %swift -module-name Swift -target x86_64-unknown-windows-msvc -dump-clang-diagnostics -I %S/Inputs -enable-cxx-interop -emit-ir %s -parse-stdlib -parse-as-library -disable-legacy-type-info | %FileCheck %s -check-prefix=MICROSOFT_X64
+
+import Constructors
+
+typealias Void = ()
+
+public func createHasVirtualBase() -> HasVirtualBase {
+  // - The `this` parameter should carry a `noalias` attribute, as it is
+  //   guaranteed that nothing will alias the object before it has been fully
+  //   constructed. It should also carry an `sret` attribute to indicate that
+  //   this is an out parameter for a structure being returned by the function.
+  //   Note that this doesn't apply on ABIs (Itanium ARM, Microsoft x64)
+  //   where we insert an (inlined) thunk; in this case, we're getting the
+  //   attributes of the constructor that was generated by Clang, which doesn't
+  //   insert these attributes.
+  //
+  // - The `this` parameter should _not_ carry a `nocapture` attribute (unlike
```

Comment:
If the type is loadable in Swift, `nocapture` is permitted, I think.

In this example, the function returns the result indirectly, and the result is marked `nocapture`. Since we construct a C++ object in that memory, the constructor must also be `nocapture`

Reply:
Hm... good point.

Since there is no way in general to guarantee that the C++ constructor will be `nocapture`, I think the conclusion here is that when we're returning C++ objects indirectly, we should always do so without `nocapture`?

(We could, in principle, add the `nocapture` back if we can prove to ourselves that the pointer can't be captured. But I think Swift currently isn't set up to do that -- I assume the place that adds the `nocapture` is in `SignatureExpansion`, and that only sees the type of the function, not its definition.)

Reply:
Well, we can probably check whether the type is trivially copyable.  If it's trivially copyable, it can't be captured, because Swift is absolutely not going to promise not to move it around.

Reply:
I think I see what you're saying. Swift can certainly also move non-trivially-copyable C++ types around as it pleases, but then the move constructor will notice this happening and can adjust the pointer that was previously captured by the constructor. But a trivially copyable type has no way of doing this, and so we can assume that the constructor doesn't capture `this`. Is this what you're getting at?

Reply:
FWIW, I've checked, and Clang doesn't add `nocapture` to the `this` parameter of a constructor, even if the type is trivially copyable. I'm not sure why this is, but if Clang isn't adding the `nocapture`, I don't think we should be adding it either.

There's a wider discussion of what we should be doing WRT to `nocapture` for Swift functions that return C++ types (see https://bugs.swift.org/browse/SR-12845), but that's outside the scope of this PR.